akka-stream/src/main/scala/akka/stream/javadsl/Source.scala

```diff
@@ -2292,6 +2292,41 @@ final class Source[Out, Mat](delegate: scaladsl.Source[Out, Mat]) extends Graph[
       case elem if clazz.isInstance(elem) => supplier.get()
     }: PartialFunction[Throwable, Graph[SourceShape[Out], NotUsed]])
 
+  /**
+   * CompleteOn allows to complete the stream when an upstream error occurs.
+   *
+   * Since the underlying failure signal onError arrives out-of-band, it might jump over existing elements.
+   * This operator can recover the failure signal, but not the skipped elements, which will be dropped.
+   *
+   * '''Emits when''' element is available from the upstream
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes or failed with exception is an instance of the provided type
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def completeOn(clazz: Class[_ <: Throwable]): javadsl.Source[Out, Mat] = completeOn(ex => clazz.isInstance(ex))
+
+  /**
+   * CompleteOn allows to complete the stream when an upstream error occurs.
+   *
+   * Since the underlying failure signal onError arrives out-of-band, it might jump over existing elements.
+   * This operator can recover the failure signal, but not the skipped elements, which will be dropped.
+   *
+   * '''Emits when''' element is available from the upstream
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes or failed with predicate return ture
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def completeOn(predicate: java.util.function.Predicate[_ >: Throwable]): javadsl.Source[Out, Mat] =
```

Comment:
I know the naming has been discussed, but this name doesn't include much about that it is about errors. Might be difficult to discover. `onErrorComplete` is better. I still like `recoverWithComplete` because it's close to what we already have, but seems like @johanandren didn't like it.

Reply:
There are :
- `onErrorComplete(..)`
- `onErrorContinue(...)`
- `onErrorMap(...)`
- `onErrorReturn(...)`
- `onErrorResume(...)`
- `onErrorStop(...)`

in reactor-core.  and should we use `completeOn` , `completeWhen`, or `recoverWithComplete` here, naming is always hard.

Reply:
Both `CompleteOn` and `CompleteWhen` seems reasonable to accept a predicate/condition, but for `recoverWithComplete` for me I think it should take zero parameters.

Reply:
@octonato What's your input? the `recoverWithComplete` align better but seems like `completeWhen` is a great name too.

Reply:
I agree that `onErrorComplete` is better for discoverability and it express much better that the method has the intention to be used in case of an error. 

About `recoverWithComplete` is less interesting, I think. Internally, we are 'recovering' with empty Source, but that's an implementation detail. 

From a user perspective, I don't think recover makes sense because we are not recovering (and continuing) when an error occurs. We are completing gracefully. 

That said, maybe a overloaded `onErrorComplete` (Predicated and PF) will do the job. Clear intention and discoverable.

Reply:
@johanandren I am going to update this PR with `onErrorComplete` , as the https://github.com/akka/akka-persistence-r2dbc is using `Flux` and in `Flux` it's named `onErrorComplete` too.

Reply:
I asked chatgpt, it suggests `completeWhen`.

Reply:
@johanandren I updated the PR to `onErrorComplete` and the https://github.com/akka/akka-persistence-r2dbc is using `Flux` too.

And as we here we only matching on error types, and `completeOn`/`completeWhen` is more generic.

Reply:
Let's go with onErrorComplete